tests/framework/framework.go

```diff
@@ -96,28 +137,107 @@ func getMaxFailsFromEnv() int {
 // NewKubernetesReporter creates a new instance of the reporter.
 func NewKubernetesReporter() *KubernetesReporter {
 	return &KubernetesReporter{
+		FailureCount: 0,
+		artifactsDir: os.Getenv("ARTIFACTS"),
+		maxFails:     getMaxFailsFromEnv(),
+	}
+}
+
+func NewFramework() *Framework {
+	f := &Framework{
 		BackupNamespace: getBackupNamespaceFromEnv(),
 		Region:          getRegionFromEnv(),
 		StorageClass:    getStorageClassFromEnv(),
-		FailureCount:    0,
-		artifactsDir:    os.Getenv("ARTIFACTS"),
-		maxFails:        getMaxFailsFromEnv(),
+		Clients:         ClientsInstance,
+		reporter:        reporter,
 	}
+
+	ginkgo.BeforeEach(f.BeforeEach)
+	ginkgo.AfterEach(f.AfterEach)
+
+	return f
 }
 
-// Dump dumps the current state of the cluster. The relevant logs are collected starting
-// from the since parameter.
-func (r *KubernetesReporter) Dump(duration time.Duration) {
+// BeforeEach provides a set of operations to run before each test
+func (f *Framework) BeforeEach() {
+	ginkgo.By("Building a namespace api object")
+	ns, err := f.CreateNamespace()
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	f.Namespace = ns
+	f.AddNamespaceToDelete(ns)
+}
+
+// AfterEach provides a set of operations to run after each test
+func (f *Framework) AfterEach() {
+	// delete the namespace(s) in a defer in case future code
+	// added here could generate an exception.
+	defer func() {
+		for _, ns := range f.namespacesToDelete {
+			defer func() { f.namespacesToDelete = nil }()
+			if ns == nil || len(ns.Name) == 0 {
+				continue
+			}
+			ginkgo.By(fmt.Sprintf("Destroying namespace %q for this suite.", ns.Name))
+			err := DeleteNS(f.K8sClient, ns.Name)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+	}()
+
+	if ginkgo.CurrentGinkgoTestDescription().Failed {
+		f.reporter.FailureCount++
+		f.reporter.Dump(f.K8sClient, f.KvClient, ginkgo.CurrentGinkgoTestDescription().Duration)
```

Comment:
might want to consider https://github.com/kubevirt/containerized-data-importer/pull/2485
for easier debugging

Reply:
nice